x/sync/network_server.go

```diff
@@ -184,33 +100,46 @@ func (s *NetworkServer) HandleChangeProofRequest(
 
 	startRoot, err := ids.ToID(req.StartRootHash)
 	if err != nil {
-		return err
+		return nil, &common.AppError{
+			Code:    p2p.ErrUnexpected.Code,
+			Message: fmt.Sprintf("failed to parse start root hash: %s", err),
+		}
 	}
 
 	endRoot, err := ids.ToID(req.EndRootHash)
 	if err != nil {
-		return err
+		return nil, &common.AppError{
+			Code:    p2p.ErrUnexpected.Code,
+			Message: fmt.Sprintf("failed to parse end root hash: %s", err),
+		}
 	}
 
 	for keyLimit > 0 {
-		changeProof, err := s.db.GetChangeProof(ctx, startRoot, endRoot, start, end, int(keyLimit))
+		changeProof, err := g.db.GetChangeProof(ctx, startRoot, endRoot, start, end, int(keyLimit))
 		if err != nil {
 			if !errors.Is(err, merkledb.ErrInsufficientHistory) {
 				// We should only fail to get a change proof if we have insufficient history.
 				// Other errors are unexpected.
-				return err
+				// TODO define custom errors
+				return nil, &common.AppError{
```

Comment:
This is because previously the network server used `AppRequest`, implementing the VM `AppHandler`. Errors are fatal so we would swallow any non-fatal errors and return nil, and return an err if the error was fatal.

In `p2p`, errors are never fatal. We return `[]byte, error` where `[]byte` is the app response, and `error` is the app error. Which ever one is set is sent back to the requesting node id.